tests/src/test/scala/common/BaseWsk.scala

```diff
@@ -0,0 +1,365 @@
+  apihost: String = WhiskProperties.getEdgeHost,
+  token: String = "") {
+  def overrides = Seq("-i", "--apihost", apihost, "--apiversion", apiversion)
+  def writeFile(propsfile: File) = {
+    val propsStr =
+      s"NAMESPACE=$namespace\nAPIVERSION=$apiversion\nAUTH=$authKey\nAPIHOST=$apihost\nAPIGW_ACCESS_TOKEN=$token\n"
+    val bw = new BufferedWriter(new FileWriter(propsfile))
+    try {
+      bw.write(propsStr)
+    } finally {
+      bw.close()
+    }
+  }
+}
+
+trait WaitFor {
+
+  /**
+   * Waits up to totalWait seconds for a 'step' to return value.
+   * Often tests call this routine immediately after starting work.
+   * Performs an initial wait before entering poll loop.
+   */
+  def waitfor[T](step: () => T,
+                 initialWait: Duration = 1 second,
+                 pollPeriod: Duration = 1 second,
+                 totalWait: Duration = 30 seconds): T = {
+    Thread.sleep(initialWait.toMillis)
+    val endTime = System.currentTimeMillis() + totalWait.toMillis
+    while (System.currentTimeMillis() < endTime) {
+      val predicate = step()
+      predicate match {
+        case (t: Boolean) if t =>
+          return predicate
+        case (t: Any) if t != null && !t.isInstanceOf[Boolean] =>
+          return predicate
+        case _ if System.currentTimeMillis() >= endTime =>
+          return predicate
+        case _ =>
+          Thread.sleep(pollPeriod.toMillis)
+      }
+    }
+    null.asInstanceOf[T]
+  }
+}
+
+trait BaseWsk extends BaseRunWsk {
+  val action: BaseAction
+  val trigger: BaseTrigger
+  val rule: BaseRule
+  val activation: BaseActivation
+  val pkg: BasePackage
+  val namespace: BaseNamespace
+  val api: BaseApi
+}
+
+trait FullyQualifiedNames {
+
+  /**
+   * Fully qualifies the name of an entity with its namespace.
+   * If the name already starts with the PATHSEP character, then
+   * it already is fully qualified. Otherwise (package name or
+   * basic entity name) it is prefixed with the namespace. The
+   * namespace is derived from the implicit whisk properties.
+   *
+   * @param name to fully qualify iff it is not already fully qualified
+   * @param wp whisk properties
+   * @return name if it is fully qualified else a name fully qualified for a namespace
+   */
+  def fqn(name: String)(implicit wp: WskProps) = {
+    val sep = "/" // Namespace.PATHSEP
+    if (name.startsWith(sep) || name.count(_ == sep(0)) == 2) name
+    else s"$sep${wp.namespace}$sep$name"
+  }
+
+  /**
+   * Resolves a namespace. If argument is defined, it takes precedence.
+   * else resolve to namespace in implicit WskProps.
+   *
+   * @param namespace an optional namespace
+   * @param wp whisk properties
+   * @return resolved namespace
+   */
+  def resolve(namespace: Option[String])(implicit wp: WskProps) = {
+    val sep = "/" // Namespace.PATHSEP
+    namespace getOrElse s"$sep${wp.namespace}"
+  }
+}
+
+trait BaseListOrGetFromCollection extends FullyQualifiedNames {
+  self: BaseRunWsk =>
```

Comment:
Do we need the extra qualification to `BaseRunWsk` @rabbah? As `self` is unused, I guess not?

Reply:
We can remove this line.